FlowCrypt/src/main/java/com/flowcrypt/email/jetpack/workmanager/HandlePasswordProtectedMsgWorker.kt

```diff
@@ -0,0 +1,424 @@
+    return rawMimeMsg
+  }
+
+  private suspend fun getMimeMessage(
+    account: AccountEntity,
+    msgEntity: MessageEntity,
+    attachments: List<AttachmentEntity>
+  ) = EmailUtil.createMimeMsg(
+    context = applicationContext,
+    sess = Session.getDefaultInstance(Properties()),
+    account = account,
+    msgEntity = msgEntity.copy(isEncrypted = false),
+    atts = attachments
+  )
+
+  private suspend fun getAttachments(msgEntity: MessageEntity) =
+    roomDatabase.attachmentDao().getAttachmentsSuspend(
+      account = msgEntity.email,
+      label = JavaEmailConstants.FOLDER_OUTBOX,
+      uid = msgEntity.uid
+    ).map {
+      it.copy(
+        forwardedFolder = "Outbox",
+        forwardedUid = Random.nextLong(),
+        decryptWhenForward = true
+      )
+    }
+
+  private suspend fun getRecipients(
+    mimeMsgWithAttachments: MimeMessage,
+    type: Message.RecipientType
+  ) = withContext(Dispatchers.IO) {
+    mimeMsgWithAttachments.getMatchingRecipients(
+      type = type,
+      list = GeneralUtil.getRecipientsWithoutUsablePubKeys(
+        context = applicationContext,
+        emails = mimeMsgWithAttachments.getAddresses(type)
+      )
+    )
+  }
+
+  private suspend fun updateExistingMimeMsgWithUrl(
+    msgEntity: MessageEntity,
+    url: String
+  ) = withContext(Dispatchers.IO) {
+    val mimeMsgWithoutAttachments = MimeMessage(
+      Session.getDefaultInstance(Properties()),
+      msgEntity.rawMessageWithoutAttachments?.toInputStream()
+    )
+    val fromAddress = (mimeMsgWithoutAttachments.from.first() as InternetAddress).address
+    val multipart = mimeMsgWithoutAttachments.content as Multipart
+    multipart.addBodyPart(MimeBodyPart().apply {
+      setText(applicationContext.getString(R.string.password_protected_msg_promo, fromAddress, url))
+    }, 0)
+    //todo-denbond7 need to add HTML version
+    mimeMsgWithoutAttachments.saveChanges()
+
+    val out = ByteArrayOutputStream()
+    mimeMsgWithoutAttachments.writeTo(out)
+
+    roomDatabase.msgDao().updateSuspend(
+      msgEntity.copy(
+        rawMessageWithoutAttachments = String(out.toByteArray()),
+        state = MessageState.QUEUED.value
+      )
+    )
+  }
+
+  private suspend fun uploadMsgToFESAndReturnUrl(
+    apiRepository: FlowcryptApiRepository,
+    domain: String,
+    idToken: String,
+    messageUploadRequest: MessageUploadRequest,
+    pwdEncryptedWithAttachments: String
+  ): String = withContext(Dispatchers.IO) {
+    val messageUploadResponseResult = apiRepository.uploadPasswordProtectedMsgToWebPortal(
+      context = applicationContext,
+      domain = domain,
+      idToken = idToken,
+      messageUploadRequest = messageUploadRequest,
+      msg = pwdEncryptedWithAttachments
+    )
+
+    if (messageUploadResponseResult.status
+      != com.flowcrypt.email.api.retrofit.response.base.Result.Status.SUCCESS
+    ) {
+      throw IllegalStateException("Uploading a password-protected message to a web portal failed")
+    }
```

Comment:
This could happen from time to time, eg network error. Would this later be retried automatically? If so, under what conditions? Or only when retried manually by user?

Could the `IllegalStateException` contain  the wrong `messageUploadResponseResult.status` so that it's clearer why it failed?

Reply:
> This could happen from time to time, eg network error. Would this later be retried automatically? If so, under what conditions? Or only when retried manually by user?

Please look at `handleExceptionsForMessage`

https://github.com/FlowCrypt/flowcrypt-android/blob/e3fc9f391ad3f264c6c548535f05190c0996af1d/FlowCrypt/src/main/java/com/flowcrypt/email/jetpack/workmanager/HandlePasswordProtectedMsgWorker.kt#L200

If it will be a network error it will restart automatically. Otherwise only manually(a user will see a notifaction about that + Outbox count on the main screen. So the user shouldn't skip it.).

> Could the IllegalStateException contain the wrong messageUploadResponseResult.status so that it's clearer why it failed?

Yes, we can add more details